pkg/storage/etcd/etcd_helper.go

```diff
@@ -289,14 +289,20 @@ func (h *etcdHelper) decodeNodeList(nodes []*etcd.Node, slicePtr interface{}) er
 	for _, node := range nodes {
 		if node.Dir {
 			trace.Step("Decoding dir " + node.Key + " START")
-			if err := h.decodeNodeList(node.Nodes, slicePtr); err != nil {
+			if err := h.decodeNodeList(node.Nodes, filter, slicePtr); err != nil {
 				return err
 			}
 			trace.Step("Decoding dir " + node.Key + " END")
 			continue
 		}
 		if obj, found := h.getFromCache(node.ModifiedIndex); found {
-			v.Set(reflect.Append(v, reflect.ValueOf(obj).Elem()))
+			if matches, err := filter(obj); err == nil {
+				if matches {
+					v.Set(reflect.Append(v, reflect.ValueOf(obj).Elem()))
+				}
+			} else {
```

Comment:
Invert this if so that it is:

```
matches, err := filter(obj)
if err != nil {
  return err
}
if matches ...
```

That's more idiomatic for Go - we don't return in else blocks anywhere else in our codebase.